OpenSim/Analyses/OutputReporter.cpp

```diff
@@ -0,0 +1,129 @@
+// ANALYSIS
+//=============================================================================
+int OutputReporter::begin(SimTK::State& s)
+{
+    if (!proceed()) return(0);
+
+    _model = _model->clone();
+
+    _tableReporterDouble = new TableReporter_<double>();
+    _tableReporterVec3 = new TableReporter_<SimTK::Vec3>();
+    _tableReporterSpatialVec = new TableReporter_<SimTK::SpatialVec>();
+
+    _model->addComponent(_tableReporterDouble.get());
+    _model->addComponent(_tableReporterVec3.get());
+    _model->addComponent(_tableReporterSpatialVec.get());
+
+    for (int i = 0; i < getProperty_output_paths().size(); ++i) {
+        auto outpath = ComponentPath(get_output_paths(i));
+        // assume that Outputs are for the Model
+        string compName = _model->getAbsolutePathString();
+        // if a a ComponentPath is specified then use the parent path
+        if (outpath.getNumPathLevels()) {
+            compName = outpath.getParentPathString();
+        }
+        auto outputName = outpath.getComponentName();
+
+        const Component* comp = _model;
+        if (!compName.empty())
+            comp = &_model->getComponent(compName);
+
+        auto& out = comp->getOutput(outputName);
+        if (out.getTypeName() == "double") {
+            _tableReporterDouble->addToReport(out);
+        }
+        else if (out.getTypeName() == "Vec3") {
+            _tableReporterVec3->addToReport(out);
+        }
+        else if (out.getTypeName() == "SpatialVec") {
+            _tableReporterSpatialVec->addToReport(out);
+        }
+        else {
+            cout << "Output '" << out.getPathName() << "' of type "
+                << out.getTypeName() << " is not supported by OutputReporter."
+                << "Consider adding a TableReporter_ to the Model." << endl;
+        }
+    }
+
+    _model->initSystem();
+
+    return 0;
+}
+
+int OutputReporter::step(const SimTK::State& s, int stepNumber)
+{
+    if (!proceed(stepNumber)) {
+        return 0;
+    }
+    if (_model == nullptr)
+        return -1;
+
+    _model->realizeReport(s);
+
+    return 0;
+        
+}
+
+
+int OutputReporter::printResults(const std::string& baseName,
+    const std::string& dir,  double dT, const std::string& extension)
+{
+    if (!getOn()) {
+        printf("OutputReporter.printResults: Off- not printing.\n");
+        return 0;
+    }
+
+    auto& tableD = _tableReporterDouble->getTable();
+    if (tableD.getNumColumns()) {
+        STOFileAdapter_<double>::write(tableD,
+            dir + baseName + extension);
+    }
+
+    auto& tableV3 = _tableReporterVec3->getTable();
+    if (tableV3.getNumColumns()) {
+        STOFileAdapter_<SimTK::Vec3>::write(tableV3,
+            dir + baseName + "Vec3" + extension);
+    }
+
+    auto& tableSV = _tableReporterSpatialVec->getTable();
+    if (tableSV.getNumColumns()) {
+        STOFileAdapter_<SimTK::SpatialVec>::write(tableSV,
+            dir + baseName + "SpatialVec" + extension);
+    }
```

Comment:
I don't think you should assume the user requested a STO file. What if `extension` is `.csv`? Instead, you could rely on the file adapter registry by using `FileAdapter::writeFile` (I think that's the name of the function).

The documentation for this class should also specify the file format used, and whether or not it depends on the extension provided by the user.

Alternatively, you could give an error message if `extension = ".sto"`.

Reply:
I think the extension option should be removed for writing the results of  any Analysis. There are no supported cases of reporting non `.sto` files from an `Analysis`. I can add that.